Gems/Atom/Feature/Common/Code/Source/CoreLights/SimpleSpotLightFeatureProcessor.cpp

```diff
@@ -141,45 +144,90 @@ namespace AZ
+                        currentIdx++;
+                    }
+
+                    m_goboArrayChanged = false;
+                }
+
                 m_lightBufferHandler.UpdateBuffer(m_lightData.GetDataVector<0>());
                 m_deviceBufferNeedsUpdate = false;
             }
 
             if (r_enablePerMeshShaderOptionFlags)
             {
-                // Helper lambdas
-                auto indexHasShadow = [&](LightHandle::IndexType index) -> bool
-                {
-                    SpotLightUtils::ShadowId shadowId = SpotLightUtils::ShadowId(m_lightData.GetData<0>(index).m_shadowIndex);
-                    return shadowId.IsValid();
-                };
+                const uint32_t lightAndShadow = m_lightMeshFlag.GetIndex() | m_shadowMeshFlag.GetIndex();
+                const uint32_t lightOnly = m_lightMeshFlag.GetIndex();
 
-                // Filter lambdas
-                auto hasShadow = [&](const MeshCommon::BoundsVariant& bounds) -> bool
-                {
-                    return indexHasShadow(m_lightData.GetIndexForData<1>(&bounds));
-                };
-                auto noShadow = [&](const MeshCommon::BoundsVariant& bounds) -> bool
+                AZStd::vector<MeshCommon::BoundsVariant> lightsWithShadow;
+                AZStd::vector<MeshCommon::BoundsVariant> lightsWithoutShadow;
+
+                for (int32_t idx = 0; idx < m_lightData.GetDataCount(); idx++)
                 {
-                    return !indexHasShadow(m_lightData.GetIndexForData<1>(&bounds));
-                };
+                    const auto& lightData = m_lightData.GetData<0>(idx);
+                    const auto& extraData = m_lightData.GetData<1>(idx);
 
-                // Mark meshes that have point lights without shadow using only the light flag.
-                MeshCommon::MarkMeshesWithFlag(GetParentScene(), AZStd::span(m_lightData.GetDataVector<1>()), m_lightMeshFlag.GetIndex(), noShadow);
+                    SpotLightUtils::ShadowId shadowId = SpotLightUtils::ShadowId(lightData.m_shadowIndex);
+                    if (shadowId.IsValid())
+                    {
+                        lightsWithShadow.push_back(extraData.m_boundsVariant);
+                    }
+                    else
+                    {
+                        lightsWithoutShadow.push_back(extraData.m_boundsVariant);
+                    }
+                }
 
-                // Mark meshes that have point lights with shadow using a combination of light and shadow flags.
-                uint32_t lightAndShadow = m_lightMeshFlag.GetIndex() | m_shadowMeshFlag.GetIndex();
-                MeshCommon::MarkMeshesWithFlag(GetParentScene(), AZStd::span(m_lightData.GetDataVector<1>()), lightAndShadow, hasShadow);
+                MeshCommon::MarkMeshesWithFlag(GetParentScene(), AZStd::span(lightsWithoutShadow), lightOnly);
+                MeshCommon::MarkMeshesWithFlag(GetParentScene(), AZStd::span(lightsWithShadow), lightAndShadow);
```

Comment:
instead of making a temporary vector for both... could you call `MeshCommon::MarkMeshesForBounds` inside the for loop?

Reply:
It would requires duplicate all the if/else checks inside the function `void MarkMeshesWithFlag(AZ::RPI::Scene* scene, AZStd::span<const BoundsVariant> boundsCollection, AZ::RPI::Cullable::FlagType flag, Filter filter = {})`.